File: experimental/greg/cppocp/OcpMultipleShooting.cpp
// OcpMultipleShooting.hpp
// Greg Horn
// Casadi 2011

#include "OcpMultipleShooting.hpp"

using namespace CasADi;
using namespace std;

OcpMultipleShooting::OcpMultipleShooting(Ode * _ode) : Ocp(_ode) {}

OcpMultipleShooting::~OcpMultipleShooting() {}

void OcpMultipleShooting::setTimeInterval(CasADi::SX _t0, CasADi::SX _tf)
{
	t0 = _t0;
	tf = _tf;

	// dynamics constraint
	SX dt = (tf - t0)/(N - 1);

	vector<SX>::const_iterator iter;

	for (int k=0; k<N-1; k++){
		SXMatrix x0 = getStateMat(k);
		SXMatrix x1 = getStateMat(k+1);
		
		SXMatrix u0 = getActionMat(k);
		SXMatrix u1 = getActionMat(k+1);

		SXMatrix p = getParamMat();

		SX tk   = k*dt;
		SX tkp1 = (k+1)*dt;
		
		SXMatrix xErr = x1 - ode->rk4Step( x0, u0, u1, p, tk, tkp1);

		addNonlconEq( xErr );
	}
}

void OcpMultipleShooting::discretize(int _N)
{
	ode->assertUnlocked();
	ode->locked = 1;

	N = _N;

	designVariables = create_symbolic("designVariables", getBigN());

	for (int k=0; k<getBigN(); k++){
		lb.push_back(-1e50);
		ub.push_back( 1e50);
		//	-numeric_limits<double>::infinity()

		guess.push_back(0);
	}
}


// total number of discretized states/actions/params
int OcpMultipleShooting::getBigN()
{
	return N*ode->nxu() + ode->np();
}

// get state at proper timestep
SX OcpMultipleShooting::getState(string x, int timeStep)
{
	return designVariables[getStateActionIdx(x, timeStep)];
}

// get action at proper timestep
SX OcpMultipleShooting::getAction(string u, int timeStep)
{
	return designVariables[getStateActionIdx(u, timeStep)];
}

// get param at proper timestep
SX OcpMultipleShooting::getParam(string p)
{
	return designVariables[getParamIdx(p)];
}

// calculate the index of states/actions at proper timestep
int OcpMultipleShooting::getStateActionIdx(string xu, int timeStep)
{
	map<string, int>::const_iterator xIter, uIter;

	xIter = ode->states.find(xu);
	uIter = ode->actions.find(xu);

	if (xIter != ode->states.end()) // state
		return timeStep*ode->nxu() + xIter->second;
	else if (uIter != ode->actions.end()) // action
		return timeStep*ode->nxu() + ode->nx() + uIter->second;
	else {
		cerr << "Error - \"" << xu << "\" not found in states/actions" << endl;
		throw -1;
	}
	return -1;
}

// calculate the index of params
int OcpMultipleShooting::getParamIdx(string p)
{
	map<string, int>::const_iterator pIter;

	pIter = ode->params.find(p);

	if (pIter != ode->params.end())
		return N*ode->nxu() + pIter->second;
	else {
		cerr << "Error - \"" << p << "\" not found in params" << endl;
		throw -1;
	}
	return -1;
}


void OcpMultipleShooting::boundStateAction(string xu, double _lb, double _ub, int timeStep)
{
	int idx = getStateActionIdx(xu, timeStep);
	lb[idx] = _lb;
	ub[idx] = _ub;
}


void OcpMultipleShooting::boundParam(string p, double _lb, double _ub)
{
	int idx = getParamIdx(p);
	lb[idx] = _lb;
	ub[idx] = _ub;
}

SXMatrix OcpMultipleShooting::getStateMat(int timeStep)
{
	SXMatrix ret = create_symbolic("a_state", ode->nx());
	// SXMatrix ret(ode->nx(), 1);
	map<string,int>::const_iterator xIter;
	for (xIter = ode->states.begin(); xIter != ode->states.end(); xIter++)
		ret[xIter->second] = getState(xIter->first, timeStep);

	return ret;
}

SXMatrix OcpMultipleShooting::getActionMat(int timeStep)
{
	SXMatrix ret = create_symbolic("an_action", ode->nu());
	//	SXMatrix ret(ode->nu(),1);
	map<string,int>::const_iterator uIter;
	for (uIter = ode->actions.begin(); uIter != ode->actions.end(); uIter++)
		ret[uIter->second] = getAction(uIter->first, timeStep);

	return ret;
}

SXMatrix OcpMultipleShooting::getParamMat()
{
	SXMatrix ret = create_symbolic("a_param", ode->np());
	// SXMatrix ret(ude->np(), 1);
	map<string,int>::const_iterator pIter;
	for (pIter = ode->params.begin(); pIter != ode->params.end(); pIter++)
		ret[pIter->second] = getParam(pIter->first);

	return ret;
}

File: experimental/greg/ocpMultipleShooting.py
#
#     This file is part of CasADi.
#
#     CasADi -- A symbolic framework for dynamic optimization.
#     Copyright (C) 2010-2014 Joel Andersson, Joris Gillis, Moritz Diehl,
#                             K.U. Leuven. All rights reserved.
#     Copyright (C) 2011-2014 Greg Horn
#
#     CasADi is free software; you can redistribute it and/or
#     modify it under the terms of the GNU Lesser General Public
#     License as published by the Free Software Foundation; either
#     version 3 of the License, or (at your option) any later version.
#
#     CasADi is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#     Lesser General Public License for more details.
#
#     You should have received a copy of the GNU Lesser General Public
#     License along with CasADi; if not, write to the Free Software
#     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
#
import ocp
import ode
import casadi as C

class OcpMultipleShooting(ocp.Ocp):

    def setTimeInterval(self, t0, tf):
        self.t0 = t0
        self.tf = tf

        # dynamics constraint
        dt = (self.tf - self.t0)/(self.N - 1)
        for k in range(self.N-1):
            x0 = self._getStateVec(k)
            x1 = self._getStateVec(k+1)

            u0 = self._getActionVec(k)
            u1 = self._getActionVec(k+1)

            p = self._getParamVec()
    
            t0 = k*dt
            t1 = (k+1)*dt
    
            xErr = x1 - self.ode.rk4Step( x0, u0, u1, p, t0, t1)
            
            #self.addNonlcon( xErr, "==", C.MX(self.ode._Nx()*[0]))
            self.addNonlcon( xErr, "==", C.SX(self.ode._Nx()*[0]))
            #self._addNonlconIneq( xErr )


    def discretize(self, N):
        if self.ode.locked:
            errStr = "Ode "+self.name+" has already been discretized and is in read-only mode"
            raise ValueError(errStr)
        self.ode.locked = True

        self.N = N

        #self.designVariables = C.MX('designVariables', self._getBigN())
        self.designVariables = C.ssym('designVariables', self._getBigN())

        self.lb = [-1e-15 for k in range(self._getBigN())]
        self.ub = [ 1e-15 for k in range(self._getBigN())]
        self.guess = [ 0 for k in range(self._getBigN())]


    def _getBigN(self): # total number of discretized states/actions/params
        return self.N*self.ode._Nxu() + self.ode._Np()

    def _getIdx(self, xup, timeStep=None):
        if isinstance(timeStep, list):
            raise TypeError("timeStep not supposed to be a list")

        # states/actions
        if xup in self.ode.states+self.ode.actions:
            if timeStep==None:
                raise ValueError("state/action timeStep == None")
            return self.ode._Nxu()*timeStep + (self.ode.states+self.ode.actions).index(xup)

        # parameters
        elif xup in self.ode.params:
            if timeStep!=None:
                raise ValueError("params don't have an associated timestep")
            return self.N*self.ode._Nxu() + self.ode.params.index(xup)

    def bound(self, xup, lb, ub, timeStep=None):
        # states/actions
        if xup in self.ode.states+self.ode.actions:
            if timeStep == None:
                timeStep = range(0, self.N)
            elif not isinstance(timeStep, list):
                timeStep = [timeStep]
            for ts in timeStep:
                idx = self._getIdx(xup, ts)
                self.lb[idx] = lb
                self.ub[idx] = ub

        # params
        elif xup in self.ode.params:
            if timeStep != None:
                errStr = "params don't have timesteps in xfull (param \""+xup+"\")"
                raise ValueError(errStr)
            idx = self._getIdx(xup)
            self.lb[idx] = lb
            self.ub[idx] = ub

        # not states/actions/params
        else:
            errStr = "bound() cannot find \""+xup+"\" in states/actions/parameters"
            raise NameError(errStr)

    def _getStateActionVec(self, timeStep):
        if timeStep >= self.N:
            errStr = "timeStep "+str(timeStep)+" is out of range (0, "+str(self.N-1)+")"
            raise ValueError(errStr)
        return self.designVariables[ timeStep*self.ode._Nxu() : (timeStep+1)*self.ode._Nxu() ]

    def _getStateVec(self, timeStep):
        xu = self._getStateActionVec(timeStep)
        return xu[ 0 : self.ode._Nx() ]

    def _getActionVec(self, timeStep):
        xu = self._getStateActionVec(timeStep)
        return xu[self.ode._Nx() : self.ode._Nx() + self.ode._Nu()]

    def _getParamVec(self):
        return self.designVariables[ self.ode._Nxu()*self.N : self.ode._Nxu()*self.N + self.ode._Np() ]

    def getParams(self):
        return self.ode._paramVecToDict(self._getParamVec())
        
    def getState(self, timeStep):
        return self.ode._stateVecToDict(self._getStateVec(timeStep))
        
    def getAction(self, timeStep):
        return self.ode._actionVecToDict(self._getActionVec(timeStep))
